syntax = "proto3";

package chat_memory;

// The memory service definition
service ChatMemoryService {
  // Searches for relevant memories and generates a chat response
  rpc ChatWithMemories (ChatRequest) returns (ChatResponse);
  
  // Adds new memories to the system
  rpc AddMemories (AddMemoriesRequest) returns (AddMemoriesResponse);
  
  // Searches for memories without generating a response
  rpc SearchMemories (SearchRequest) returns (SearchResponse);
}

// The request message for chat with memories
message ChatRequest {
  string message = 1;
  string user_id = 2;
  repeated string source_ids = 3;
  string context = 4;
}

// The response message for chat
message ChatResponse {
  string response = 1;
  repeated string relevant_memories = 2;
  bool success = 3;
  string error = 4;
}

// The request message for adding memories
message AddMemoriesRequest {
  repeated ChatMessage messages = 1;
  string user_id = 2;
}

// The response message for adding memories
message AddMemoriesResponse {
  bool success = 1;
  string error = 2;
}

// The request message for searching memories
message SearchRequest {
  string query = 1;
  string user_id = 2;
  int32 limit = 3;
}

// The response message for searching memories
message SearchResponse {
  repeated string memories = 1;
  bool success = 2;
  string error = 3;
}

// Chat message structure
message ChatMessage {
  string role = 1;
  string content = 2;
}
